rust: implement `PointerWrapper` for unit type.

wedsonaf opened the pull request:
This is useful for cases when we don't need to store any information as
context. Implementing `PointerWrapper` for `()` allows us to directly
use it as the context type.

Signed-off-by: Wedson Almeida Filho <wedsonaf@google.com>

bjorn3 commented on rust/kernel/types.rs:
```diff
@@ -144,6 +144,22 @@ impl<T> PointerWrapper for *mut T {
+    type Borrowed<'a> = ();
+
+    fn into_pointer(self) -> *const c_types::c_void {
+        0usize as _
```
Maybe use 1 instead? There may be places where C code checks against a null pointer. In addition it is less dangerous when accidentally turning this pointer into a reference. By the way this code can also be written as `core::ptr::null()`.

wedsonaf replied:
Good idea, I'll change this.

bjorn3 commented:
> as opposed to, for example, Box<()>, which would require an allocation.

Box and Vec don't allocate for ZST's. Instead they cast an integer to a pointer. (the alignment of the type to be precise. this gives a non-null aligned pointer)

wedsonaf commented:
> > as opposed to, for example, Box<()>, which would require an allocation.
> 
> Box and Vec don't allocate for ZST's. Instead they cast an integer to a pointer. (the alignment of the type to be precise. this gives a non-null aligned pointer)

This is interesting. So if we have two instances of Box<T> where T is a ZST, we can't tell them apart?

bjorn3 commented:
Correct.

wedsonaf commented:
PTAL.

* Updated commit message to remove claim that `Box<()>` allocates memory.
* Return `1` as the pointer value now.
* Fixed clippy errors (omitting `()` as the return value).

wedsonaf commented:
> Correct.

I need to review the `LockedBy` code to make sure all pointer comparisons there are expected to work then.
